builder/frameworks/uf2conv.py

```diff
@@ -318,30 +346,77 @@ def dev_uploader(target, source, env):
     global appstartaddr
     appstartaddr = int(env.address, 0)
     bin_name = join(env.get("BUILD_DIR"), env.get("PROGNAME"))+'.bin'
+    elf_name = join(env.get("BUILD_DIR"), env.get("PROGNAME"))+'.elf'
     uf2_name = join(env.get("BUILD_DIR"), env.get("PROGNAME"))+'.uf2'
     drive = env.get("UPLOAD_PORT")
     if None != env.GetProjectOption("monitor_port"):
-        try: # reset usb stdio
-            usb = serial.Serial( env.GetProjectOption("monitor_port"), 1200)
+        try:  # reset usb stdio
+            usb = serial.Serial(env.GetProjectOption("monitor_port"), 1200)
             time.sleep(0.1)
             usb.close()
         except:
             pass
-        time.sleep(1.0) # Windows - AutoPlay
-        if 'Windows' not in system(): time.sleep(1.0)
-    print( "  Converting to UF2 ( 0x%x )" % (appstartaddr) )
-    with open( bin_name, mode='rb' ) as f: inpbuf = f.read()
+        time.sleep(1.0)  # Windows - AutoPlay
+        if 'Windows' not in system():
+            time.sleep(1.0)
+    print("  Converting to UF2 ( 0x%x )" % (appstartaddr))
+    with open(bin_name, mode='rb') as f:
+        inpbuf = f.read()
     outbuf = convert_to_uf2(inpbuf)
     time.sleep(.1)
-    write_file(uf2_name, outbuf) # write uf2 to build folder
+    write_file(uf2_name, outbuf)  # write uf2 to build folder
+
+    # Uploads via picoprobe protocol
+    if env.GetProjectOption("upload_protocol") == "picoprobe":
+        # TODO: tidy this part
+
+        # Gets Picoprobe configuration
+        picoprobe = env.BoardConfig().__dict__[
+            "_manifest"]["debug"]["tools"]["picoprobe"]
+        
+        # path to picoprobe binary
+        path = join(env.PioPlatform().get_package_dir(
+            'tool-pico-openocd'), picoprobe["server"]["executable"])
+        
+        # extra arguments to configure openocd
+        arguments = ' '.join(picoprobe["server"]["arguments"]).replace("$PACKAGE_DIR", env.PioPlatform().get_package_dir(

```

Comment:
best practice is to keep each command parameter in its own string in a list.

```suggestion
        ocd_package_dir = env.PioPlatform().get_package_dir('tool-pico-openocd')
        arguments = [arg.replace("$PACKAGE_DIR", ocd_package_dir) for arg in picoprobe["server"]["arguments"]]
```

Joining everything into one big string will, at best, cause issues when someone uses a space in a filename. At worst, it'll allow for execution of arbitrary programs and is a security risk.